File: 2016/day02-2.hs
import System.IO

type Position = (Int, Int)

solution :: String -> String
solution content = map toCode . map (followInstructions (0, 2)) . concatInstructions [] . lines $ content

concatInstructions :: [String] -> [String] -> [String]
concatInstructions xs [] = xs
concatInstructions [] (y:ys) = concatInstructions [y] ys
concatInstructions xs (y:ys) = concatInstructions (xs ++ [last xs ++ y]) ys

toCode :: Position -> Char
toCode (2,0) = '1'
toCode (1,1) = '2'
toCode (2,1) = '3'
toCode (3,1) = '4'
toCode (0,2) = '5'
toCode (1,2) = '6'
toCode (2,2) = '7'
toCode (3,2) = '8'
toCode (4,2) = '9'
toCode (1,3) = 'A'
toCode (2,3) = 'B'
toCode (3,3) = 'C'
toCode (2,4) = 'C'
toCode _ = ' '

followInstructions :: Position -> String -> Position
followInstructions p [] = p
followInstructions p (x:xs) = followInstructions (follow x p) xs

follow :: Char -> Position -> Position
follow 'U' (0,2) = (0,2)
follow 'U' (1,1) = (1,1)
follow 'U' (2,0) = (2,0)
follow 'U' (3,1) = (3,1)
follow 'U' (4,2) = (4,2)
follow 'U' (x,y) = (x, y-1)
follow 'D' (0,2) = (0,2)
follow 'D' (1,3) = (1,3)
follow 'D' (2,4) = (2,4)
follow 'D' (3,3) = (3,3)
follow 'D' (4,2) = (4,2)
follow 'D' (x,y) = (x, y+1)
follow 'L' (2,0) = (2,0)
follow 'L' (1,1) = (1,1)
follow 'L' (0,2) = (0,2)
follow 'L' (1,3) = (1,3)
follow 'L' (2,4) = (2,4)
follow 'L' (x,y) = (x-1, y)
follow 'R' (2,0) = (2,0)
follow 'R' (3,1) = (3,1)
follow 'R' (4,2) = (4,2)
follow 'R' (3,3) = (3,3)
follow 'R' (2,4) = (2,4)
follow 'R' (x,y) = (x+1, y)

follow _ current = current

main :: IO ()
main = do
    handle <- openFile "day02.dat" ReadMode
    content <- hGetContents handle
    print . solution $ content

File: 2016/day02-2.ts
type Position = [number, number];

function solution(content: string) {
  const lines = content.split("\n");
  let current: Position = [0, 2];
  let code = "";
  lines.forEach((line) => {
    line.split("").forEach((char) => {
      current = follow(char, current);
    });
    code += toCode(current);
  });
  return code;
}

function toCode(position: Position) {
  if (position[0] === 2 && position[1] === 0) return "1";
  if (position[0] === 1 && position[1] === 1) return "2";
  if (position[0] === 2 && position[1] === 1) return "3";
  if (position[0] === 3 && position[1] === 1) return "4";
  if (position[0] === 0 && position[1] === 2) return "5";
  if (position[0] === 1 && position[1] === 2) return "6";
  if (position[0] === 2 && position[1] === 2) return "7";
  if (position[0] === 3 && position[1] === 2) return "8";
  if (position[0] === 4 && position[1] === 2) return "9";
  if (position[0] === 1 && position[1] === 3) return "A";
  if (position[0] === 2 && position[1] === 3) return "B";
  if (position[0] === 3 && position[1] === 3) return "C";
  if (position[0] === 2 && position[1] === 4) return "D";
  return "";
}

function follow(instruction: string, current: Position): Position {
  if (isAllowed("U", instruction, current)) {
    return [current[0], current[1] - 1];
  }
  if (isAllowed("D", instruction, current)) {
    return [current[0], current[1] + 1];
  }
  if (isAllowed("L", instruction, current)) {
    return [current[0] - 1, current[1]];
  }
  if (isAllowed("R", instruction, current)) {
    return [current[0] + 1, current[1]];
  }
  return current;
}

function isAllowed(direction: string, instruction: string, current: Position) {
  if (instruction !== direction) return false;
  if (direction === "U")
    return !["0-2", "1-1", "2-0", "3-1", "4-2"].includes(hash(current));
  if (direction === "D")
    return !["0-2", "1-3", "2-4", "3-3", "4-2"].includes(hash(current));
  if (direction === "L")
    return !["2-0", "1-1", "0-2", "1-3", "2-4"].includes(hash(current));
  if (direction === "R")
    return !["2-0", "3-1", "4-2", "3-3", "2-4"].includes(hash(current));
  return false;
}

function hash(position: Position) {
  return `${position[0]}-${position[1]}`;
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
